CN0503 Platform Fluorescence Decay Support

NSagan271 opened the pull request:
Add support for four fluorescence-decay-related commands in the CN0503 firmware (in which, the decay of a fluorophore is measured over a period of time after an initial LED pulse.) See commit [249690f](https://github.com/analogdevicesinc/EVAL-ADICUP3029/commit/249690fda19f024deef3264191273237089468e7) for a lengthy description of the command functionality, as well as helper functions implemented from each command.

In gpio.c, add a function to register an interrupt handler to a GPIO pin. This allows the fluorescence decay measurements to use an interrupt-based FIFO sampling method.

In flash.c, fix a few bugs regarding the use of the flash address (as far as I can tell, the flash address was used to represent a byte index in some places, and a word index in others).

adrimbarean commented on projects/ADuCM3029_demo_cn0503/src/platform_source/flash.c:
```diff
@@ -112,7 +112,7 @@ static int32_t flash_write_page(struct flash_dev *dev, uint32_t flash_addr,
 
 	for(i = 0; i < array_size; i++)
-		temp_ptr[(i + (flash_addr & 0x7FF))] = array[i];
+		temp_ptr[(i + (flash_addr & 0x7FF) / 4)] = array[i];
```
Curious, can you describe a bit what sort of bugs you encountered here?

NSagan271 replied:
Flash writes were hanging for addresses that were not page-aligned. First, the while loop in flash_write was an infinite loop because temp (the amount by which the data index was being incremented) got set to 0. Then, it would hang or error on adi_fee_GetPageNumber in some cases. Not sure why that was happening, but being careful about whether flash_addr referred to bytes or words fixed it.

adrimbarean commented on projects/ADuCM3029_demo_cn0503/src/cn0503.c:
```diff
@@ -3307,66 +3588,1420 @@ int32_t cn0503_prompt(struct cn0503_dev *dev)
-	float a, b;
-	int8_t i = 0, op_id;
+	struct cn0503_impulse_response *impresp = dev->impulse_response;
+	uint16_t integ_offset_lower, integ_offset_upper, i = impresp->data_cycle;
```
nitpick: is there a need here to use i instead of impresp->data_cycle directly?

NSagan271 replied:
No, this was laziness. I can change i to impresp->data_cycle to avoid allocating more stack space.

adrimbarean commented:
I think the [last commit](https://github.com/analogdevicesinc/EVAL-ADICUP3029/pull/88/commits/1849d991ae99d50f74353d7ea866ad0cc2e75c79) can be merged with the previous one and a force push done.

adrimbarean reviewed (changes requested):
We need to make sure the license from the [Matlib file](https://github.com/analogdevicesinc/EVAL-ADICUP3029/pull/88/files#diff-428a990f0e19daac177df91f58ba9c24e0e7d48235c06b46ed18358a9dd7df0e) is compatible with ours. Standing by until it's clear.

NSagan271 commented:
> We need to make sure the license from the [Matlib file](https://github.com/analogdevicesinc/EVAL-ADICUP3029/pull/88/files#diff-428a990f0e19daac177df91f58ba9c24e0e7d48235c06b46ed18358a9dd7df0e) is compatible with ours. Standing by until it's clear.

The license for the library is [here](https://www.codeproject.com/info/cpol10.aspx). If it isn't compatible with ours, I do not have to use it. I'm only doing 2-by-2 matrix operations, which are reasonable to implement myself.

NSagan271 commented:
Updated PR, addressing all comments.
**Notes**:
- I removed the Matlib library; it ended up being simpler to implement to needed functionality myself.
- I needed to modify one line in irq.c for setting the interrupt polarity in order to get interrupts to work the proper way.
- Is there any way to re-run the checks? It seems like Azure failed to allocate resources for the checks, so they were cancelled.

mchindri commented:
> Updated PR, addressing all comments.
> **Notes**:
> 
> * I removed the Matlib library; it ended up being simpler to implement to needed functionality myself.
> * I needed to modify one line in irq.c for setting the interrupt polarity in order to get interrupts to work.
> * Is there any way to re-run the checks? It seems like Azure failed to allocate resources for the checks, so they were cancelled.

There is an issue on azure for the pipelines. 
https://status.dev.azure.com/
When it is fixed, the checks can be retriggered

adrimbarean commented on projects/ADuCM3029_demo_cn0503/src/adpd410x.c:
```diff
@@ -569,18 +568,18 @@ int32_t adpd410x_read_fifo(struct adpd410x_dev *dev, uint32_t *data,
-		// Can read a maximum of 255 bytes at once
-		next_packet_size = (total_bytes - bytes_read > 255) ? 255 : total_bytes -
-				   bytes_read;
+		next_packet_size = total_bytes - bytes_read;
+		// Can read a maximum of 255 bytes at once for i2c
+		if (dev->dev_type == ADPD4101 && next_packet_size > 255)
+			next_packet_size = 255;
```
Why is this change needed?

adrimbarean replied:
Never mind, I can see the difference. Wrote comment 1 second before I realized.

adrimbarean commented on projects/ADuCM3029_demo_cn0503/src/adpd410x.c:
```diff
@@ -76,11 +76,11 @@ int32_t adpd410x_reg_read(struct adpd410x_dev *dev, uint16_t address,
 
 /**
- * @brief Read a single byte from device register.
+ * @brief Read a specified number of bytes from device register.
```
Can we have these changes in a separate commit so we can propagate them later to the no-OS repository?

adrimbarean commented on projects/ADuCM3029_demo_cn0503/src/platform_source/flash.c:
```diff
@@ -148,7 +148,7 @@ int32_t flash_write(struct flash_dev *dev, uint32_t flash_addr,
 
 	do {
-		temp = (FLASH_PAGE_SIZE_BYTES / 4) - (flash_addr & 0x7FF);
+		temp = (FLASH_PAGE_SIZE_BYTES - (flash_addr & 0x7FF)) / 4;
```
Can we have these changes in a separate commit as well? It would be much easier to integrate them outside of the project if needed.

adrimbarean commented on projects/ADuCM3029_demo_cn0503/src/platform_source/irq.c:
```diff
@@ -235,7 +235,8 @@ int32_t irq_register_callback(struct irq_ctrl_desc *desc, uint32_t irq_id,
 		adi_gpio_GetGroupInterruptPolarity(gpio_port, &gpio_pin);
 		gpio_pin &= ~BIT((gpio_desc->number & 0xF));
-		gpio_pin |= BIT(aducm_desc->conf[irq_id].gpio_conf->mode);
+		gpio_pin |= aducm_desc->conf[irq_id].gpio_conf->mode <<
+			    (gpio_desc->number & 0xF);
```
Separate commit if possible, please.

adrimbarean reviewed (approved):
This looks good to me! Just a few comments I feel will help with the busy work after, but I will approve now.

NSagan271 commented:
Moved changes to driver files to separate commits.
